File: code_library/n_queen.c
#include <stdio.h>

int n, lim, counter;

void backtrack(int i, int c, int l, int r){
    if (!i){
        counter++;
        return;
    }

    int bitmask, x;
    --i, bitmask = lim & ~(l | r | c);

    while (bitmask){
        x = (-bitmask & bitmask);
        if (!x) return;
        bitmask ^= x;
        backtrack(i, c | x, (l | x) << 1, (r | x) >> 1);
    }
}

int count_ways(int dimension){
	n = dimension;
	counter = 0, lim = (1 << n) - 1;
	backtrack(n, 0, 0, 0);

	return counter;
}

int main(){
    printf("%d\n", count_ways(8));   /// 92
    printf("%d\n", count_ways(13));  /// 73712
    printf("%d\n", count_ways(15));  /// 2279184

    return 0;
}

File: code_library/python/n_queen.py
def count_ways(n):
    """
    Parameters
        ----------
        n : Dimension of chessboard (n x n)

    Returns
        ----------
        The number of ways to place n queens on the chessboard,
        such that no pair of queens attack each other
    """

    def backtrack(i, c, l, r):
        if not i:
            return 1

        ways, mask = 0, ((1 << n) - 1) & ~(l | r | c)
        while mask:
            x = -mask & mask
            mask ^= x
            ways += backtrack(i - 1, c | x, (l | x) << 1, (r | x) >> 1)

        return ways

    return backtrack(n, 0, 0, 0)


def main():
    assert count_ways(8) == 92
    assert count_ways(13) == 73712


if __name__ == '__main__':
    main()
